feature: add uncertainty info to info dict

sankha555 opened the pull request:
* [x] related to #133 
* [x] add support to include uncertainties info dict in reports which are implemented by the ```distrib``` package
* [x] tests added/passed

sankha555 commented:
@matthiaskoenig As a first step of the patch, I have added some code in the ```info_base``` to make a dict for the uncertainties. Please take a look at the patch and let me know if this is something that we are looking for and what changes I should make to it. I have extracted information for the uncertainties and their nested uncertainty parameters, but am not sure how to represent them as the HTML code form. For the time being, I have extracted them and stored them as dicts. 

I haven't added any explicit tests to check for the uncertainty examples, but I extracted the examples from the test data and ran the regular tests for report generation on them and they are passing those. Once we fully implement the uncertainties dicts and rendering, I would add some possible uncertainty specific tests to test the feature. Kindly take a look and let me know what additions I should make and please let me know how I should proceed with the rendering part( like any specific format for the HTML fragmnents and the HTML template file, etc.). Thanks!

matthiaskoenig commented on src/sbmlutils/report/sbmlinfo.py:
```diff
@@ -156,6 +156,43 @@ def info_sbase(sbase: libsbml.SBase) -> Dict[str, Any]:
                 info["replaced_elements"] = replaced_elements_combined
 
+        # distrib
+        item_distib = sbase.getPlugin("distrib")
```
typo, rename to `sbase_distrib`

matthiaskoenig commented on src/sbmlutils/report/sbmlinfo.py:
```diff
@@ -156,6 +156,43 @@ def info_sbase(sbase: libsbml.SBase) -> Dict[str, Any]:
+            listOfUncertainties = item_distib.getListOfUncertainties()
+            for uncertainty in listOfUncertainties:
+
+                u_dict = {
```
The uncertainty is an SBase, so you have to get the sbase info first, i.e. the SBase dictionary. See similar code for the other objects.

matthiaskoenig replied:
see `info_sbaseref`

matthiaskoenig commented on src/sbmlutils/test/report/test_report.py:
```diff
@@ -11,10 +11,27 @@
 )
 
 
+uncertainty_paths = [
```
I assume these are the models from the sbml test suite? We should add the complete testsuite models if these are not too big, not only the distrib models.

matthiaskoenig replied:
You have to add the models as part of the pull request so the tests run in CI.

sankha555 replied:
I took a look at all the models under the ```data``` directory, they aren't too voluminous and I think we can add all of them to the tests. I am thinking of creating a new directory ```testsuite``` under ```test/data``` and copying and renaming each model from the ```data``` directory(further partitioned inside some more directories) to a naming scheme that can be accessed using a for loop. Would this approach be suitable for the task Sir?

matthiaskoenig replied:
Please keep the existing models as is. You could add a list for them in the `test/init.py` which you can then use in other tests. Yes add a `test/data/testsuite` for the testsuite.

matthiaskoenig commented on src/sbmlutils/test/__init__.py:
```diff
@@ -31,6 +32,8 @@
 
 REPRESSILATOR_SBML = MODELS_DIR / "repressilator" / "BIOMD0000000012_urn.xml"
 
+UNCERTAINTY_PATH = TEST_DIR / "data" / "distrib" / "testsuite"
```
Add the complete testsuite if not too big. Use 
```
TESTSUITE_PATH = TEST_DIR / "data" / "testsuite"
```

matthiaskoenig commented on src/sbmlutils/test/report/test_report.py:
```diff
@@ -11,10 +11,27 @@
 
 
+uncertainty_paths = [
+    UNCERTAINTY_PATH / f"00{i:0>3}-sbml-l3v1.xml"
```
some duplication: First get the uncertainty_model_ids:
```
uncertainty_model_ids = [i for i in range(40, 48)]
        + [49, 50, 51, 52, 56, 65, 69]
        + [i for i in range(69, 104)]

uncertainty_paths = []
for level_version in ["l3v1", "l3v2"]:
    for i in uncertainty_model_ids:
         uncertainty_paths.append(
             TESTSUITE_PATH / f"00{i:0>3}-sbml-{level_version}.xml"
          )  
```
As a rule of thumb list and dictionary comprehensions should never span more then a single line.
Otherwise a for loop is much more readable.

matthiaskoenig commented on src/sbmlutils/test/report/test_report.py:
```diff
@@ -69,3 +86,9 @@ def check_report_math_type(sbml_path: Path, math_type: str, tmp_path):
     assert isinstance(html, str)
+
+
+@pytest.mark.parametrize("sbml_path", uncertainty_paths, ids=sbml_paths_idfn)
```
In addition to this general tests write some specific tests using the `uncertainty_example.xml` from the examples. 
- create a copy in the test data folder.
- get the info for the uncertainties and check via asserts that the content of the dictionaries is correct.

sankha555 replied:
Similar to the comment above, I would copy the ```uncertainty_example.xml``` to the new ```testsuite``` directory and run a specific test for it. I hope this would work fine for our tests. Please let me know if I should use some other approach for making the testsuite. I would do all the required changes and re-upload for your reviews. Thanks!

matthiaskoenig replied:
For the testsuite I meant to include the actual SBML testsuite models, but keep these separate from our own additional test models.

matthiaskoenig commented:
This looks already very promising. Please make the requested changes.
Best Matthias

sankha555 commented:
@matthiaskoenig Thanks for the reviews. I have taken a note of the required changes and am working on them. I have a few doubts regarding some, I have mentioned them under the corresponding comment. Thanks!

sankha555 commented:
@matthiaskoenig I have added the above changes to the patch. Please take a look at it and please let me know if I should make any more improvements to them. I have added the sbase info for the uncertainty by calling the ```info_sbase()```method using a static call-by-class approach. 

 I have also added a new ```testsuite``` directory containing all the files for the uncertainties tests. I have also added a specific test for the ```uncertainty.xml``` example in which I have fetched the information of the first uncertainty found and the first uncertainty parameter found in the uncertainty and asserted some values of their attributes to the expected values. Please let me know if this method of adding the testsuite and tests is suitable for the issue and I would make any changes needed to improve it.

I was thinking about the ```test_report.py``` file's organization a bit while making the patch. I think that the file imports and declares a significant number of model files, paths and restructures them into lists. This takes around ~70 lines of code after which the actual parameterized tests and helper functions start. Could we do a refactoring of the test file by separating the imports in the ```__init__.py``` file of the reports test module and keep just the tests in the ```test_report.py```. Do you think this would be necessary and if so shall I raise an issue and linked PR for it?

Please let me know your feedback about the patch changes. I would make the required changes to improve it upload again for review. Thanks!

matthiaskoenig commented:
Yes, please move all the test collection to the __init__.py. I.e. generate the various lists there and reuse them.

matthiaskoenig commented on src/sbmlutils/report/sbmlinfo.py:
```diff
@@ -156,6 +156,41 @@ def info_sbase(sbase: libsbml.SBase) -> Dict[str, Any]:
+            # list to store uncertainities
+            info["uncertainties"] = []
+
+            listOfUncertainties = sbml_distrib.getListOfUncertainties()
```
variables in python should be with underscore no camelcase, i.e. `list_of_uncertainties`. Better just directly iterate over the list. I.e. 

```
for uncertainty in sbml_distrib.getListOfUncertainties():
```

matthiaskoenig commented on src/sbmlutils/report/sbmlinfo.py:
```diff
@@ -156,6 +156,41 @@ def info_sbase(sbase: libsbml.SBase) -> Dict[str, Any]:
+                u_dict["uncert_params"] = []
+
+                # uncertainty parameters defined for the uncertainty
+                uncertainty_params = uncertainty.getListOfUncertParameters()
```
Same here, directly iterate over the list without generating another local variable.

matthiaskoenig reviewed (changes requested):
Make the minor changes and then I can merge the pull request.

sankha555 commented:
@matthiaskoenig I removed the redundant local variables. Also, removing the ```type(sbml_distrib) == libsbml.DistribSBasePlugin``` test fails the tests as it then goes on to find uncertainties for the ```DistribSBMLDocumentPlugin```class too which has doesn't have methods like ```getListOfUncertainties()```. I hence corrected the test to use ```isinstance()``` for the check. 

I would create a new issue for refactoring the ```test_report.py``` and link a PR to it. Please let me know if I should make any more changes. Thanks!

matthiaskoenig commented:
@sankha555 Please have a look at the changes I did in https://github.com/matthiaskoenig/sbmlutils/pull/217/commits/dcaca57d76ab00704d926d167ab9e4ae183d5405;

- naming of the keys must match the SBML object names
- add type annotations for clarity
- fix test structure: clear separation in distrib test and core report tests
- some of the tests were incorrect and pmathml was not tested
- removed some duplicate tests
- paths collected in test/__init__.py

sankha555 commented:
@matthiaskoenig I have taken a note of all these changes. I will make sure to follow these structures and guidelines in adding tests for the future pull requests. Thanks a lot for the help!
